tutorial/20_recipes/013_wilcoxon_pruner.py

```diff
@@ -0,0 +1,324 @@
+and operations research, and has numerous practical applications in logistics,
+manufacturing, and DNA sequencing, among others.
+The problem is classified as NP-hard, so approximation algorithms or
+heuristic methods are commonly employed for larger instances.
+
+One simple heuristic method applicable to TSP is simulated annealing (SA).
+SA starts with an initial solution (it can be constructed by a simpler heuristic
+like greedy method), and it randomly checks the neighborhood (defined later)
+of the solution. If a neighbor is better, the solution is updated to the neighbor.
+If the neighbor is worse, SA still updates the solution to the neighbor with
+probability $e^{-\Delta c / T}$, where $\Delta c (> 0)$ is the difference of
+the cost (sum of the distance) between the new solution and the old one and
+$T$ is a parameter called "temperature". The temperature controls
+how much worsening of the solution is tolerated to escape from the local minimum
+(high means more tolerant). If the temperature is too low, SA will quickly
+fall into a local minimum; if the temperature is too high, SA will be like
+a random walk and the optimization will be inefficient. Typically, we set a
+"temperature schedule" that starts from a high temperature and gradually
+decreases to zero.
+
+There are several ways to define neighborhood for TSP, but we use a
+simple neighborhood called 2-opt. 2-opt neighbor chooses a path in
+the current solution and reverses the visiting order in the path.
+For example, if the initial solution is `a→b→c→d→e→a`, `a→d→c→c→e→a` is
+a 2-opt neighbor (the path from `b` to `d` is reversed).
+This neighborhood is good because computing the difference of the cost
+can be done in constant time (we only need to care about the start
+and the end of the chosen path).
+
+Main Tutorial: Tuning SA Parameters for TSP
+====================================================
+"""
+
+from dataclasses import dataclass
+import math
+
+import numpy as np
+import optuna
+import plotly.graph_objects as go
+from numpy.linalg import norm
+
+
+@dataclass
+class SAOptions:
+    max_iter: int = 10000
+    T0: float = 1.0
+    alpha: float = 2.0
+    patience: int = 50
+
+
+def tsp_cost(vertices: np.ndarray, idxs: np.ndarray) -> float:
+    return norm(vertices[idxs] - vertices[np.roll(idxs, 1)], axis=-1).sum()
+
+
+###################################################################################################
+# Greedy solution for initial guess.
+
+
+def tsp_greedy(vertices: np.ndarray) -> np.ndarray:
+    idxs = [0]
+    for _ in range(len(vertices) - 1):
+        dists_from_last = norm(vertices[idxs[-1], None] - vertices, axis=-1)
+        dists_from_last[idxs] = np.inf
+        idxs.append(np.argmin(dists_from_last))
+    return np.array(idxs)
+
+
+###################################################################################################
+# .. note::
+#     The following `tsp_simulated_annealing` function can be acceralated by `numba`.
```

Comment:
```suggestion
```
How about removing this line because simply adding `@numba.njit` will result in error?